dotnet/G3DViewer/MainViewModel.cs

```diff
@@ -250,35 +262,177 @@ internal int AddG3DData(Ara3D.IG3D g3dFile)
             Ara3D.IArray<int> faceSizes = Ara3D.G3DExtensions.FaceSizes(g3dFile);
 
             var vertexData = g3dFile.VertexAttribute.ToVector3s();
-            var indexData = g3dFile.IndexAttribute.ToInts();
+            var uvData = Ara3D.G3DExtensions.UVs(g3dFile);
+            var indexData = Ara3D.G3DExtensions.CornerVertexIndices(g3dFile);
 
+            float aabbMinX = float.MaxValue;
+            float aabbMaxX = float.MinValue;
+            float aabbMinY = float.MaxValue;
+            float aabbMaxY = float.MinValue;
+            float aabbMinZ = float.MaxValue;
+            float aabbMaxZ = float.MinValue;
 
-            var b1 = new MeshBuilder(true, true, true);
-            var l1 = new LineBuilder();
+            Vector3Collection sharpDxVectos = new Vector3Collection(vertexData.Count);
+            Vector2Collection sharpDxUVs = new Vector2Collection(vertexData.Count);
 
+            displayStats.NumVertices += vertexData.Count;
+            displayStats.NumFaces += faceCount;
+
+            for (int vertexIndex = 0; vertexIndex < vertexData.Count; vertexIndex++)
+            {
+                sharpDxVectos.Add(Ara3DToSharpDX(vertexData[vertexIndex]));
+                sharpDxUVs.Add(Ara3DToSharpDX(uvData[vertexIndex]));
+
+                aabbMinX = Math.Min(aabbMinX, vertexData[vertexIndex].X);
+                aabbMinY = Math.Min(aabbMinY, vertexData[vertexIndex].Z);
+                aabbMinZ = Math.Min(aabbMinZ, vertexData[vertexIndex].Y);
+
+                aabbMaxX = Math.Max(aabbMaxX, vertexData[vertexIndex].X);
+                aabbMaxY = Math.Max(aabbMaxY, vertexData[vertexIndex].Z);
+                aabbMaxZ = Math.Max(aabbMaxZ, vertexData[vertexIndex].Y);
+            }
+
+            displayStats.AABB = new Ara3D.Box(new Ara3D.Vector3(aabbMinX, aabbMinY, aabbMinZ), new Ara3D.Vector3(aabbMaxX, aabbMaxY, aabbMaxZ));
+
+            int currentFace = 0;
             int globalVectorIndex = 0;
-            for (int face = 0; face < faceCount; face++)
+
+            while (currentFace < faceCount)
             {
-                int faceSize = faceSizes[face];
+                var geometry3d = new MeshGeometry3D();
+                var triangleIndices = new IntCollection(100000 * 3);

```

Comment:
Where does that magic number come from? Try to always use named constants. What happens if the number of faces is greater than that number? It looks like it quietly cuts them out. Instead it show throw a big old exception so that developers know that this code is unfinished.

Reply:
Meshes are split into smaller ones for rendering, this is the sub-mesh size. Will change to a constant.